sphinx/application.py

```diff
@@ -310,23 +311,34 @@ def _init_env(self, freshenv):
                     logger.info(_('failed: %s'), err)
                 self._init_env(freshenv=True)
 
-    def _init_builder(self, buildername):
-        # type: (unicode) -> None
+    def create_builder(self, buildername):
+        # type: (unicode) -> Builder
         if buildername is None:
             logger.info(_('No builder selected, using default: html'))
             buildername = 'html'
         if buildername not in self.builderclasses:
             raise SphinxError(_('Builder name %s not registered') % buildername)
 
         builderclass = self.builderclasses[buildername]
-        self.builder = builderclass(self)
+        return builderclass(self)
+
+    def _init_builder(self):
+        # type: () -> None
+        self.builder.set_environment(self.env)
+        self.builder.init()
```

Comment:
Right. We have already many 3rd party builders.
In my PR, I don't change the interface of `Builder.init()`, because I was worry about to break them.
+0 for giving `env` to init function, but we also have to care the old interfaces like following:
```
try:
    # new interface of init() here
    self.builder.init(self.env)
except TypeError:
    # for old extensions
    self.builder.set_environment(self.env)
    self.init()
```

Either is fine for me.